### PR TITLE
Include logic to retrieve indexImage and bundleImage for CSV

### PR DESCRIPTION
**Pull request by ramperher**
This change is required for DCI in order to:

- Define indexImage and bundleImage in the Operator type
- Include the logic needed to retrieve these values and save them in the new variables created
- Provide these variables in the claim.json

With this, we can run CNF Cert Suite on DCI and, based on the claim.json file, we can retrieve all the data from the operators tested to feed preflight test suite.

### CODE REVIEW

File: pkg/config/configsections/common.go [hunk 1]
```diff
@@ -39,6 +39,12 @@ type Operator struct {
 	// Subscription name is required field, Name of used subscription.
 	SubscriptionName string `yaml:"subscriptionName" json:"subscriptionName"`
 
+	// BundleImage is the URL referencing the bundle image
+	BundleImage string `yaml:"bundleImage" json:"bundleImage"`
```
Comment (edcdavid) [inline pkg/config/configsections/common.go:43]: Created a PR (on your PR/ main) adding saving the whole initial runtime environment, including this structure to the claims.jon file. It also includes some linter fixes.
https://github.com/ramperher/test-network-function/pull/1

Comment (ramperher) [reply pkg/config/configsections/common.go:43]: thanks David! I've added your changes, I'll start checking how this change works by running tnf on DCI, checking if we have the desired output. I'll keep you all posted.

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 2]
```diff
@@ -260,6 +263,31 @@ func getConfiguredOperatorTests() []string {
 	return opTests
 }
 
+// setBundleAndIndexImage provides the bundle image and index image for a given CSV.
+// These variables are saved in the `configsections.Operator` in order to be used by DCI,
+// which obtains them from the claim.json and provides them to preflight suite.
+func setBundleAndIndexImage(csvName, csvNamespace string) (bundleImage, indexImage string) {
```
Comment (sebrandon1) [inline pkg/config/autodiscover/autodiscover_targets.go:269]: Could we have unit tests to cover this function?

**Review by sebrandon1 — COMMENTED** [review]
Just one request for unit tests for a new function.

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 2 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go:269]: Sure, I'll take a look, I'm waiting for having some resources ready in our labs to do some testing

**Comment by ramperher** [comment]
Hi guys, I have been doing some manual checks for this, because I did not manage to have a successful DCI job with example-cnf, but I can confirm it works when testing example-cnf (at least with some operators that were installed fine during the deployment).

This is what I've done so far:

1. Try to run example-cnf on our lab, it failed, but some operators were installed correctly, so that I did the tests with them. I used trex-operator for this example.
2. I set the autodiscovery labels for the operator, as we do here: https://github.com/dci-labs/example-cnf-config/blob/master/testpmd/hooks/roles/apply-cnf-cert-config/tasks/install.yml#L39-L52
3. I configured tnf_config.yaml file, just including example-cnf under targetNameSpaces section
4. I configured env variables for tnf execution:

```
export TNF_NON_INTRUSIVE_ONLY=true
export TNF_RUN_CFD_TEST=false
export LOG_LEVEL=debug
export TNF_PARTNER_REPO=registry_name:registry_port/testnetworkfunction
# this last one, we generated the version based on this PR
export TNF_IMAGE=test-network-function:hashId
```

6. Then I run the tnf container

```
./run-tnf-container.sh -k /path/to/kubeconfig -i $TNF_IMAGE -t /path/to/test-network-function -o /path/to/test-network-function -f operator
```

7. Operator tests were passed, but that's not the important thing here. If we check the generated claim.json file... then surprise:

```
        "operators": [
          {
            "Org": "nfv-example-cnf-catalog",
            "Version": "v0.2.9",
            "bundleImage": "\"quay.io/rh-nfv-int/trex-operator-bundle@sha256:12b0cfe22acd3ff4b2c2f1497e4068e06b44e0ca71d0519f131d5f4158e03e82\"",
            "indexImage": "\"quay.io/rh-nfv-int/nfv-example-cnf-catalog@sha256:7060cee796af17543fdd448dcace211abd532c186bb042cf4057c5213ed16c01\"",
            "name": "trex-operator.v0.2.9",
            "namespace": "example-cnf",
            "packag": "trex-operator.v0.2.9",
            "subscriptionName": "lb-operator-subscription",
            "tests": [
              "OPERATOR_STATUS"
            ]
          }
        ],
{code}

The only thing that I don't like is that " " are embedded in both indexImage and bundleImage. I'm not a Go expert, but I guess this can be fixed easily. I will check

I will try to set up some unit tests to check this " " issue and then try to have a full DCI job with the claim.json there.

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 2 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go:269]: @sebrandon1 , now I remember that I am calling this method when isTest variable is set to false (https://github.com/test-network-function/test-network-function/pull/612/files#diff-25fc9547da0d403308471a07947597646a4691b31fbecda834a8d157183da470L240), just for simplicity. So, what do you recommend for doing unit tests for this function?

Comment (sebrandon1) [reply pkg/config/autodiscover/autodiscover_targets.go:269]: Sure thing, so in `pkg/config/autodiscover/autodiscover_targets_test.go` we should have a function called `TestSetBundleAndIndexImage()` that has some expected inputs and outputs and tests normal operation plus edge cases.

Something like [this](https://github.com/test-network-function/test-network-function/blob/main/test-network-function/accesscontrol/suite_test.go#L31-L70) unit test from the accesscontrol package for example.

**Comment by edcdavid** [comment]
I added a PR to your branch for the comment below:
>The only thing that I don't like is that " " are embedded in both indexImage and bundleImage. I'm not a Go expert, but I guess this >can be fixed easily. I will check

https://github.com/ramperher/test-network-function/pull/2

**Comment by ramperher** [comment]
Unit test has been added. Still needing to have a successful run on DCI before confirming this patch works. Working on this.

**Comment by ramperher** [comment]
I'll resume this work in the following days after fixing some other stuff in our side, now we are ready to test this again with more depth. I'll keep you posted.

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 3]
```diff
@@ -260,6 +263,31 @@ func getConfiguredOperatorTests() []string {
...
+	installPlan := execCommandOutput(installPlanCmd)
+
+	// Then, retrieve the bundle image using the installplan
+	bundleImageCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].path", csvNamespace, installPlan)
+	bundleImage = strings.Replace(execCommandOutput(bundleImageCmd), "\"", "", -1)
+
+	// To retrieve the index image, we firstly need the catalogsource and the namespace in which it is deployed
+	catalogSourceNameCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].catalogSourceRef.name", csvNamespace, installPlan)
+	catalogSourceNamespaceCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].catalogSourceRef.namespace", csvNamespace, installPlan)
+	catalogSourceName := execCommandOutput(catalogSourceNameCmd)
```
[the remark this answers is not in the document]
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: Done, thanks for this advice!

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 4]
```diff
@@ -260,6 +263,31 @@ func getConfiguredOperatorTests() []string {
 	return opTests
 }
 
+// setBundleAndIndexImage provides the bundle image and index image for a given CSV.
+// These variables are saved in the `configsections.Operator` in order to be used by DCI,
+// which obtains them from the claim.json and provides them to preflight suite.
+func setBundleAndIndexImage(csvName, csvNamespace string) (bundleImage, indexImage string) {
+	// First step is to extract the installplan related to the csv
+	installPlanCmd := fmt.Sprintf("oc get installplan -n %s | grep %q | awk '{ print $1 }'", csvNamespace, csvName)
+	installPlan := execCommandOutput(installPlanCmd)
```
[the remark this answers is not in the document]
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go:335]: Done

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 5]
```diff
@@ -260,6 +263,31 @@ func getConfiguredOperatorTests() []string {
...
+
+	// Then, retrieve the bundle image using the installplan
+	bundleImageCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].path", csvNamespace, installPlan)
+	bundleImage = strings.Replace(execCommandOutput(bundleImageCmd), "\"", "", -1)
```
[the remark this answers is not in the document]
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: I've finally removed this to use go-template, but applying this advice to retrieve the indexImage.
Regarding .status.bundleLookups, it seems to be an arraw with length 1 always, so it should not be a problem.

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 6]
```diff
@@ -260,6 +263,31 @@ func getConfiguredOperatorTests() []string {
...
+
+	// To retrieve the index image, we firstly need the catalogsource and the namespace in which it is deployed
+	catalogSourceNameCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].catalogSourceRef.name", csvNamespace, installPlan)
+	catalogSourceNamespaceCmd := fmt.Sprintf("oc get installplan -n %s %s -o json | jq .status.bundleLookups[].catalogSourceRef.namespace", csvNamespace, installPlan)
+	catalogSourceName := execCommandOutput(catalogSourceNameCmd)
+	catalogSourceNamespace := execCommandOutput(catalogSourceNamespaceCmd)
+
+	// Then, we can retrieve the index image
+	indexImageCmd := fmt.Sprintf("oc get catalogsource -n %s %s -o json | jq .spec.image", catalogSourceNamespace, catalogSourceName)
+	indexImage = strings.Replace(execCommandOutput(indexImageCmd), "\"", "", -1)
```
[the remark this answers is not in the document]
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: Done

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: Done

**Comment by ramperher** [comment]
> * SUCCESS https://www.distributed-ci.io/jobs/4e8fac65-21eb-4b04-98cc-b9ff305cc256/jobStates
>   from [Include simple-demo-operator in tnf_test_example dci-labs/dallas-pipelines#267](https://github.com/dci-labs/dallas-pipelines/pull/267)

Hi again, I've tested the last changes, taking into account the suggestions from @greyerof, and it's working as expected. If you check the claim.json file generated with the tnf execution with DCI (https://www.distributed-ci.io/files/df080370-2651-4b1a-9805-c38401ff7c2f), you will see the information from the operator is extracted correctly:

```
...
        "operators": [
          {
            "Org": "simple-demo-operator-catalog",
            "Version": "v0.0.3",
            "bundleImage": "quay.io/telcoci/simple-demo-operator-bundle@sha256:e631c8dc7ab2d1ff67bf2731fd73820cb8e0214f9970d6ab1afca2f089d2fdb9",
            "indexImage": "registry.dfwt5g.lab:4443/testnetworkfunction/simple-demo-operator-catalog-disconnected:0.0.3",
            "name": "simple-demo-operator.v0.0.3",
            "namespace": "test-cnf",
            "packag": "simple-demo-operator.v0.0.3",
            "subscriptionName": "simple-demo-operator-subscription",
            "tests": [
              "OPERATOR_STATUS"
            ]
          }
        ],
...
```

So, this is ready for review again.

**Comment by ramperher** [comment]
All checks have passed after making some changes.

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 7]
```diff
@@ -323,6 +326,37 @@ func getConfiguredOperatorTests() []string {
 	return opTests
 }
 
+// getBundleAndIndexImage provides the bundle image and index image for a given CSV.
+// These variables are saved in the `configsections.Operator` in order to be used by DCI,
+// which obtains them from the claim.json and provides them to preflight suite.
+func getBundleAndIndexImage(csvName, csvNamespace string) (bundleImage, indexImage string) {
```
Comment (greyerof) [inline pkg/config/autodiscover/autodiscover_targets.go:332]: Add an error return value:
`(bundleImage, indexImage string, err error)`

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 8]
```diff
@@ -323,6 +326,37 @@ func getConfiguredOperatorTests() []string {
...
+		bundleImage = infoFromInstallPlan[0]
+		catalogSourceName := infoFromInstallPlan[1]
+		catalogSourceNamespace := infoFromInstallPlan[2]
+
+		// Then, we can retrieve the index image
+		indexImageCmd := fmt.Sprintf("oc get catalogsource -n %s %s -o json | jq -r .spec.image", catalogSourceNamespace, catalogSourceName)
+		indexImage = execCommandOutput(indexImageCmd)
+	}
+
+	return bundleImage, indexImage
```
Comment (greyerof) [inline pkg/config/autodiscover/autodiscover_targets.go:357]: `return bundleImage, indexImage, nil`

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 9]
```diff
@@ -323,6 +326,37 @@ func getConfiguredOperatorTests() []string {
...
+func getBundleAndIndexImage(csvName, csvNamespace string) (bundleImage, indexImage string) {
+	// First step is to extract the installplan related to the csv
+	installPlanCmd := fmt.Sprintf("oc get installplan -n %s | grep %q | awk '{ print $1 }'", csvNamespace, csvName)
+	installPlan := execCommandOutput(installPlanCmd)
+
+	// If no installPlan is obtained, just return an empty string for the two returned variables
+	if installPlan == "" {
+		bundleImage = ""
+		indexImage = ""
+	} else {
```
Comment (greyerof) [inline pkg/config/autodiscover/autodiscover_targets.go:341]: Fast return here, creating a custom error:
```
	if installPlan == "" {
        return "", "",  errors.New("installplan not found")
    }
```
After this fast return, you don't need the "else" branch, which also saves one indentation level.

File: pkg/config/autodiscover/autodiscover_targets.go [hunk 10]
```diff
@@ -301,6 +303,7 @@ func buildOperatorFromCSVResource(csv *CSVResource, istest bool) (op configsecti
 		op.SubscriptionName = subscriptionName[0]
 	}
 	if !istest {
+		op.BundleImage, op.IndexImage = getBundleAndIndexImage(op.Name, op.Namespace)
```
Comment (greyerof) [inline pkg/config/autodiscover/autodiscover_targets.go:306]: Check error variable right after this call:
```
op.BundleImage, op.IndexImage, err = getBundleAndIndexImage(op.Name, op.Namespace)
if err != nil {
    log.Errorf("Failed to get operator bundle and index image for csv % (ns %s), error: %s", op.Name, op.Namespace, err)
}
```

**Review by greyerof — COMMENTED** [review]
Just some comments to add error handling.

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 10 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go:306]: done

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 9 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: done

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 8 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: done

File: pkg/config/autodiscover/autodiscover_targets.go (hunk 7 above)
Comment (ramperher) [reply pkg/config/autodiscover/autodiscover_targets.go]: done

**Comment by ramperher** [comment]
> Just some comments to add error handling.

Thanks for the review @greyerof ! I've changed the code taking into account your proposal and retesting again, and everything is fine :)

**Review by greyerof — APPROVED** [review]

**Review by edcdavid — APPROVED** [review]
